### PR TITLE
Add 'delay' into Network

### PR DESCRIPTION
**Pull request by Fullstop000**
The feature could be used later in the benchmark.

Signed-off-by: Fullstop000 <fullstop1005@gmail.com>

### CODE REVIEW

File: harness/src/network.rs [hunk 1]
```diff
@@ -57,6 +57,8 @@ pub struct Network {
     pub storage: HashMap<u64, MemStorage>,
     /// Drop messages from `from` to `to` at a rate of `f64`.
     dropm: HashMap<Connection, f64>,
+    /// Delay sending messages from `from` to `to` at a rate of `f64` by blocking `u64` microseconds.
+    delaym: HashMap<Connection, (f64, u64)>,
```
Comment (Hoverbear) [inline harness/src/network.rs]: Do you think it's worth using https://doc.rust-lang.org/std/time/struct.Duration.html here?

**Comment by Hoverbear** [comment]
Oh my gosh this is such a great idea @Fullstop000 :)

File: harness/src/network.rs (hunk 1 above)
Comment (Fullstop000) [reply harness/src/network.rs]: It should be a Duration. Thanks!

**Comment by Fullstop000** [comment]
I'm thinking that current designing is single-thread based which might be unfriendly to benchmarks. For example, we have a raft cluster with nodes A, B, and C where A is the leader and B&C is a follower. If a msg from A to B is delayed, the msgs from A to C should be unaffected.

**Comment by Hoverbear** [comment]
@Fullstop000 I think that's a very valid concern. Do you think we could find a design that would permit that?

**Comment by Fullstop000** [comment]
@Hoverbear I'll try to enhance `Network` to achieve that if it is truly needed in the benchmark. The current design is suitable for most cases.

File: harness/src/network.rs [hunk 2]
```diff
@@ -57,6 +57,8 @@ pub struct Network {
     pub storage: HashMap<u64, MemStorage>,
     /// Drop messages from `from` to `to` at a rate of `f64`.
     dropm: HashMap<Connection, f64>,
+    /// Delay sending messages from `from` to `to` at a rate of `f64` by blocking `Duration` time.
+    delaym: HashMap<Connection, (f64, Duration)>,
```
Comment (BusyJay) [inline harness/src/network.rs]: This is a little confusing. So it does not just delay the messages in the connection, it also delays all the pending messages that may dispatch in other connection?

Comment (Fullstop000) [reply harness/src/network.rs]: That's the same concern I described above that connections shouldn't affect each other. But our `Network` seems to be single-thread based now but works fine in most functional test cases.

If I want to add some benches such as benchmarking proposing, this implementation is unsuitable. The design might be reconsidered and could be a lot of work.

**Comment by Fullstop000** [comment]
Close this because the feature is not suitable for current `Network` implementation. I'll try to implement a new design out of `Network` ( or maybe just enhance it ) for benchmarking.

**Pull request closed without merging**